File: packages/shared/src/api/fiveMinuteReading.ts
import { getApiClient } from "./client";
import type { OneMinuteBeat, OneMinutePendingQuestion, OneMinuteText } from "./oneMinuteReading";

/**
 * "Your Chart in Five Minutes" — docs/FIVE_MINUTE_READING_SPEC_2026-08-11.md.
 *
 * Route verified against the backend decorator before wiring:
 * `@router.get("/charts/{chart_id}/five-minute")` in app/api/charts.py — path
 * param (not query), GET, optional `asOf` query alias, same shape as
 * `getOneMinuteReading`. Two wrappers in this package have silently drifted
 * from their routes in the past, so the check is part of adding one.
 *
 * SCOPE, RIGHT NOW (updated 2026-08-11): the backend builds this for the
 * "self" register (11 beats) and "client_with_guardian" (6). `parent` and
 * `other` 404, identically to the flag being off, and by design rather than
 * pending work — see spec §0.2.
 *
 * BEAT IDS ARE DELIBERATELY NOT A UNION TYPE HERE, and that is not laziness.
 * The set has changed three times since this wrapper was written — most
 * recently gaining `the_tension`, `window_ahead` and `what_comes_after` in the
 * descent rebuild (spec §8) — and each beat is a self-describing
 * `{id, text, basis}` that a client renders generically. A union would turn
 * every backend beat addition into a breaking change across three packages for
 * a value no consumer switches on exhaustively. `web/components/
 * dashboard-five-minute-reading.tsx` keys on exactly three ids (`who_you_are`,
 * `what_this_rests_on`, `one_thing`) for chrome, and falls through for the
 * rest — which is the pattern to follow.
 */

export interface FiveMinuteReadingData {
  chartId: string;
  birthProfileId: string;
  displayName: string;
  asOf: string;
  readingWindow: { from: string; to: string };
  age: number;
  stage: string;
  ageBand: OneMinuteText;
  focusTopic: string;
  /** Only "self" is ever returned today — see module doc comment. */
  addressedTo: string;
  beats: OneMinuteBeat[];
  pendingQuestion: OneMinutePendingQuestion | null;
  wordCount: { ta: number; en: number };
  nextStep: { label: OneMinuteText; href: string };
}

export function getFiveMinuteReading(
  chartId: string,
  options?: { asOf?: string },
): Promise<{ success: boolean; data: FiveMinuteReadingData }> {
  const query = options?.asOf ? `?asOf=${encodeURIComponent(options.asOf)}` : "";
  return getApiClient().get(`/charts/${chartId}/five-minute${query}`) as Promise<{
    success: boolean;
    data: FiveMinuteReadingData;
  }>;
}

File: app/schemas/five_minute_reading.py
"""Response shapes for "Your Chart in Five Minutes" (docs/FIVE_MINUTE_READING_SPEC_2026-08-11.md).

Reuses ``OneMinuteText``/``OneMinuteBeat``/``OneMinuteQuestionOption``/
``OneMinutePendingQuestion``/``OneMinuteReadingWindow``/``OneMinuteWordCount``/
``OneMinuteNextStep``/``OneMinuteMeta`` from ``app.schemas.one_minute_reading``
rather than duplicating them — none of those shapes carry anything specific to
the two-minute reading, and the spec's own framing ("not new content") applies
to the wire shape as much as to the copy. ``FiveMinuteReadingData``/
``FiveMinuteReadingResponse`` stay their own types rather than aliasing
``OneMinuteReadingResponse`` — the two routes are separate API contracts that
may diverge (CLAUDE.md's coordination rule), and aliasing would silently
couple them.
"""
from __future__ import annotations

from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.one_minute_reading import (
    OneMinuteBeat,
    OneMinuteNextStep,
    OneMinutePendingQuestion,
    OneMinuteReadingWindow,
    OneMinuteText,
    OneMinuteWordCount,
)

__all__ = [
    "FiveMinuteMeta",
    "FiveMinuteReadingData",
    "FiveMinuteReadingResponse",
]


class FiveMinuteMeta(BaseModel):
    calculation_version: str = Field(alias="calculationVersion")
    generated_at: datetime = Field(alias="generatedAt")

    model_config = ConfigDict(populate_by_name=True)


class FiveMinuteReadingData(BaseModel):
    chart_id: UUID = Field(alias="chartId")
    birth_profile_id: UUID = Field(alias="birthProfileId")
    display_name: str = Field(alias="displayName")
    as_of: date = Field(alias="asOf")
    reading_window: OneMinuteReadingWindow = Field(alias="readingWindow")
    age: int
    stage: str
    age_band: OneMinuteText = Field(alias="ageBand")
    focus_topic: str = Field(alias="focusTopic")
    addressed_to: str = Field(alias="addressedTo")
    beats: list[OneMinuteBeat]
    pending_question: OneMinutePendingQuestion | None = Field(default=None, alias="pendingQuestion")
    word_count: OneMinuteWordCount = Field(alias="wordCount")
    next_step: OneMinuteNextStep = Field(alias="nextStep")

    model_config = ConfigDict(populate_by_name=True)


class FiveMinuteReadingResponse(BaseModel):
    success: bool = True
    data: FiveMinuteReadingData
    meta: FiveMinuteMeta

    model_config = ConfigDict(populate_by_name=True)
